app/gui/src/presenter/searcher/component_browser/provider.rs

```diff
@@ -383,26 +139,28 @@ impl Component {
     /// Initialize the [`Component`] provider, setting up proper connections to feed
     /// [grid](component_list_panel::grid::View) with entries and headers models.
     pub fn provide_new_list(
-        provider: controller::searcher::ComponentsProvider,
+        list: &Rc<component::List>,
         grid: &component_list_panel::grid::View,
     ) -> Self {
         let network = frp::Network::new("presenter::searcher::provider::Component");
 
+        // We capture weak reference to list, because searcher controllers sometimes calls
+        // [`Rc::get_mut`] on it and we want to keep it from cloning entire list.
```

Comment:
It seems fragile to allow the controller to have a side-effect of breaking this reference. We should never need to rebuild the list more than once per frame, right? Is cloning it expensive enough to justify this complexity?

Reply:
The list will often contain most (if not all) entries from the Suggestion Database, so the clone time may be significant.

And a broken reference means that the list is outdated, and there is no reason the provider should provide outdated entries. Moreover, this structure is "aware" of such side effects - the downgrading of the Rc is not obscured.